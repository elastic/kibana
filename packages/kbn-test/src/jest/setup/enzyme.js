/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * and the Server Side Public License, v 1; you may not use this file except in
 * compliance with, at your election, the Elastic License or the Server Side
 * Public License, v 1.
 */

import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });
