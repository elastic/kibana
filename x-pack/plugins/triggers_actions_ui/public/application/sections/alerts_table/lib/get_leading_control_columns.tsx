/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
import React from 'react';
import { i18n } from '@kbn/i18n';
import {
  EuiButtonIcon,
  EuiDataGridCellValueElementProps,
  EuiFlexGroup,
  EuiFlexItem,
  EuiIcon,
  EuiToolTip,
} from '@elastic/eui';
import { AlertsTableProps } from '../../../../types';

const ACTIONS_LABEL = i18n.translate(
  'xpack.triggersActionsUI.sections.alertsTable.column.actions',
  {
    defaultMessage: 'Actions',
  }
);

const VIEW_DETAILS_LABEL = i18n.translate(
  'xpack.triggersActionsUI.sections.alertsTable.leadingControl.viewDetails',
  {
    defaultMessage: 'View details',
  }
);

export const getLeadingControlColumns = ({
  leadingControlColumns,
  flyoutAlertIndex,
  setFlyoutAlertIndex,
}: {
  leadingControlColumns: AlertsTableProps['leadingControlColumns'];
  flyoutAlertIndex: number;
  setFlyoutAlertIndex: (index: number) => void;
}) => {
  return [
    {
      id: 'expand',
      width: 50,
      headerCellRender: () => {
        return <span>{ACTIONS_LABEL}</span>;
      },
      rowCellRender: (cveProps: EuiDataGridCellValueElementProps) => {
        const { visibleRowIndex } = cveProps as EuiDataGridCellValueElementProps & {
          visibleRowIndex: number;
        };
        return (
          <EuiFlexGroup gutterSize="none" responsive={false}>
            {flyoutAlertIndex === visibleRowIndex ? (
              <EuiFlexItem grow={false}>
                <EuiIcon type="alert" />
              </EuiFlexItem>
            ) : null}
            <EuiFlexItem grow={false}>
              <EuiToolTip content={VIEW_DETAILS_LABEL}>
                <EuiButtonIcon
                  size="s"
                  iconType="expand"
                  color="text"
                  onClick={() => {
                    setFlyoutAlertIndex(visibleRowIndex);
                  }}
                  data-test-subj="openFlyoutButton"
                  aria-label={VIEW_DETAILS_LABEL}
                />
              </EuiToolTip>
            </EuiFlexItem>
          </EuiFlexGroup>
        );
      },
    },
    ...leadingControlColumns,
  ];
};
