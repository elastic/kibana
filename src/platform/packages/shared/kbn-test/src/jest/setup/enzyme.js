/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the "Elastic License
 * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
 * Public License v 1"; you may not use this file except in compliance with, at
 * your election, the "Elastic License 2.0", the "GNU Affero General Public
 * License v3.0 only", or the "Server Side Public License, v 1".
 */

import { configure } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

configure({ adapter: new Adapter() });

/* eslint-env jest */

/**
 * This is a workaround to fix snapshot serialization of emotion classes when rendering React@18 using `import { render } from 'enzyme'`
 * With React@18 emotion uses `useInsertionEffect` to insert styles into the DOM, which enzyme `render` does not trigger because it is using ReactDomServer.renderToString.
 * With React@17 emotion fell back to sync cb, so it was working with enzyme `render`.
 * Without those styles in DOM the custom snapshot serializer is not able to replace the emotion class names.
 * This workaround ensures a fake emotion style tag is present in the DOM before rendering the component with enzyme making the snapshot serializer work.
 */
function mockEnsureEmotionStyleTag() {
  if (!document.head.querySelector('style[data-emotion]')) {
    const style = document.createElement('style');
    style.setAttribute('data-emotion', 'css');
    document.head.appendChild(style);
  }
}

jest.mock('enzyme', () => {
  const actual = jest.requireActual('enzyme');
  return {
    ...actual,
    render: (node, options) => {
      mockEnsureEmotionStyleTag();
      return actual.render(node, options);
    },
  };
});
