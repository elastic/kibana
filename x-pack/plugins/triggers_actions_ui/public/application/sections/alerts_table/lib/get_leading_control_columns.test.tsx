/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
import React from 'react';
import { shallow } from 'enzyme';
import { getLeadingControlColumns } from './get_leading_control_columns';

describe('getLeadingControlColumns()', () => {
  it('should return at least the flyout action control', () => {
    const setFlyoutAlertIndex = jest.fn();
    const columns = getLeadingControlColumns({
      leadingControlColumns: [],
      flyoutAlertIndex: 0,
      setFlyoutAlertIndex,
    });
    expect(columns.length).toBe(1);
    expect(columns[0].id).toBe('expand');

    const HeaderComponent = columns[0].headerCellRender;
    const header = shallow(<HeaderComponent />);
    expect(header.text()).toBe('Actions');

    const CellComponent = columns[0].rowCellRender;
    // TODO: there is an issue with types from EUI where it's not including `visibleRowIndex` (https://github.com/elastic/eui/issues/5811)
    // @ts-ignore
    const cell = shallow(<CellComponent visibleRowIndex={0} />);
    expect(cell).toMatchSnapshot();
  });

  it('should only render an icon for the the active row', () => {
    const setFlyoutAlertIndex = jest.fn();
    const columns = getLeadingControlColumns({
      leadingControlColumns: [],
      flyoutAlertIndex: 0,
      setFlyoutAlertIndex,
    });

    const CellComponent = columns[0].rowCellRender;
    // TODO: there is an issue with types from EUI where it's not including `visibleRowIndex` (https://github.com/elastic/eui/issues/5811)
    // @ts-ignore
    const cell = shallow(<CellComponent visibleRowIndex={1} />);
    expect(cell.exists('EuiIcon')).toBe(false);
  });

  it('should set the new index when the alert icon is clicked', () => {
    const setFlyoutAlertIndex = jest.fn();
    const columns = getLeadingControlColumns({
      leadingControlColumns: [],
      flyoutAlertIndex: 2,
      setFlyoutAlertIndex,
    });

    const CellComponent = columns[0].rowCellRender;
    // TODO: there is an issue with types from EUI where it's not including `visibleRowIndex` (https://github.com/elastic/eui/issues/5811)
    // @ts-ignore
    const cell = shallow(<CellComponent visibleRowIndex={2} />);
    cell.find('EuiButtonIcon').simulate('click');
    expect(setFlyoutAlertIndex).toHaveBeenCalledWith(2);
  });

  it('should render other leading controls', () => {
    const setFlyoutAlertIndex = jest.fn();
    const columns = getLeadingControlColumns({
      leadingControlColumns: [
        {
          id: 'selection',
          width: 67,
          headerCellRender: () => <span>Test header</span>,
          rowCellRender: () => <h2>Test cell</h2>,
        },
      ],
      flyoutAlertIndex: 0,
      setFlyoutAlertIndex,
    });
    expect(columns.length).toBe(2);

    const HeaderComponent = columns[1].headerCellRender;
    const header = shallow(<HeaderComponent />);
    expect(header.text()).toBe('Test header');

    const CellComponent = columns[1].rowCellRender;
    // TODO: there is an issue with types from EUI where it's not including `visibleRowIndex` (https://github.com/elastic/eui/issues/5811)
    // @ts-ignore
    const cell = shallow(<CellComponent visibleRowIndex={0} />);
    expect(cell.text()).toBe('Test cell');
  });
});
