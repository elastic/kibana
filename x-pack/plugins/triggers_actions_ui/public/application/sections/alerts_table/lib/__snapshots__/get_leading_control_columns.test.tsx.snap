// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getLeadingControlColumns() should return at least the flyout action control 1`] = `
<EuiFlexGroup
  gutterSize="none"
  responsive={false}
>
  <EuiFlexItem
    grow={false}
  >
    <EuiIcon
      type="alert"
    />
  </EuiFlexItem>
  <EuiFlexItem
    grow={false}
  >
    <EuiToolTip
      content="View details"
      delay="regular"
      display="inlineBlock"
      position="top"
    >
      <EuiButtonIcon
        aria-label="View details"
        color="text"
        data-test-subj="openFlyoutButton"
        iconType="expand"
        onClick={[Function]}
        size="s"
      />
    </EuiToolTip>
  </EuiFlexItem>
</EuiFlexGroup>
`;
